### PR TITLE
local sixpack

### PR DESCRIPTION
**Pull request by 4x4notfound**
- [x] Waiting on pypi release for sixpack-py
- [x] Add bucket param to participate once https://github.com/seatgeek/sixpack/pull/170 is merged
- [ ] Merge once an up stream pull request to seatgeek for [**Sixpack-client==1.1.2**](https://github.com/waveaccounting/sixpack-py/pull/2) is merged

### CODE REVIEW

File: djsixpack/djsixpack.py [hunk 1]
```diff
@@ -92,8 +107,31 @@ def convert(self, kpi=None):
...
+                try:
+                    participant = SixpackParticipant.objects.get(unique_attr=self.client_id, experiment_name=experiment_name)
+                    participant.convert = True
+                    participant.save()
+                except SixpackParticipant.DoesNotExist:
+                    pass
+
         except RequestException as e:
-            logger.exception("Error while trying to .convert: {err}".format(err=e))
+            logger.exception("Error while trying to .convert: %s" % e)
```
Comment (dlanger) [inline djsixpack/djsixpack.py]: ?

Comment (4x4notfound) [inline djsixpack/djsixpack.py]: Trying to comply with landscape. I'm going to revert and ignore.

Comment (dlanger) [inline djsixpack/djsixpack.py]: Interesting - reading the comment in landscape, I agree with it. What it's saying, though, is that you should do `logger.exception("Error while trying to .convert: %s", e)`. 

Both the original and this change unconditionally do the substitution, but doing it like I have here only does the substitution if the logging is actually needed, for a minor performance improvement.

**Comment by 4x4notfound** [comment]
v1.1.1 of sixpack-py will be tagged Monday

File: README.rst [hunk 2]
```diff
@@ -87,6 +87,18 @@ If something ever goes wrong - a request times out, the ``sixpack`` server disap
 calls will return the control alternative, and all ``convert`` calls will seem successful (and we'll note this happend
 in the log).
 
+-----------------
+Tracking Locally
+-----------------
+
+By passing in the argument `local` in the SixpackTest constructor you can telldjango-sixpack to create a convertable db record for each participant.
```
Comment (squishybear-zz) [inline README.rst]: Missing a space after `tell`.

Comment (squishybear-zz) [inline README.rst]: `convertible`

File: djsixpack/djsixpack.py [hunk 3]
```diff
@@ -5,6 +5,8 @@
 from django.conf import settings
 from requests.exceptions import RequestException
 
+from .models import SixpackParticipant
```
Comment (squishybear-zz) [inline djsixpack/djsixpack.py]: As per "Writing Idiomatic Python", we prefer using absolute imports to relative imports.

File: djsixpack/djsixpack.py [hunk 4]
```diff
@@ -92,8 +107,17 @@ def convert(self, kpi=None):
         experiment_name = self._get_experiment_name()
         try:
             resp = session.convert(experiment_name)
+
+            if self.local:
+                try:
+                    participant = SixpackParticipant.objects.get(unique_attr=self.client_id, experiment_name=experiment_name)
+                    participant.convert = True
```
Comment (squishybear-zz) [inline djsixpack/djsixpack.py]: Move lines 114 and 115 into an `else`.

File: djsixpack/models.py [hunk 5]
```diff
@@ -1 +1,13 @@
-# No models
+from django.db import models
+
+
+class SixpackParticipant(models.Model):
+    """
+    A model that contains fields relating to a sixpack experiment participant.
+    """
```
Comment (squishybear-zz) [inline djsixpack/models.py]: The """ that ends a multiline docstring should preferably be preceded by a blank line (https://github.com/waveaccounting/documentation/wiki/Coding-Standards#doc-strings).

File: README.rst [hunk 6]
```diff
@@ -87,6 +87,24 @@ If something ever goes wrong - a request times out, the ``sixpack`` server disap
 calls will return the control alternative, and all ``convert`` calls will seem successful (and we'll note this happend
 in the log).
 
+-----------------
+Tracking Locally
+-----------------
+
+By passing in the argument `local` in the SixpackTest constructor you can tell django-sixpack to create a convertible db record for each participant.
```
Comment (dlanger) [inline README.rst:94]: Can you add docs about why you'd want to do this/use cases for this? I get that it's creating a record in the DB, but not why I'd like to do that.

File: djsixpack/__init__.py [hunk 7]
```diff
@@ -1 +1 @@
-__version__ = "1.0.5"
+__version__ = "2.0.0"
```
Comment (dlanger) [inline djsixpack/__init__.py]: Let's just go to `1.1.0` - from [semver.org](https://semver.org):

> Given a version number MAJOR.MINOR.PATCH, increment the:
> 
> MAJOR version when you make incompatible API changes,
> MINOR version when you add functionality in a backwards-compatible manner, and
> PATCH version when you make backwards-compatible bug fixes.

File: djsixpack/models.py [hunk 8]
```diff
@@ -1 +1,14 @@
-# No models
+from django.db import models
+
+
+class SixpackParticipant(models.Model):
+    """
+    A model that contains fields relating to a sixpack experiment participant.
```
Comment (dlanger) [inline djsixpack/models.py]: Can you either make a more fulsome docstring (what do each field mean?), or drop this.

File: setup.py [hunk 9]
```diff
@@ -1,6 +1,6 @@
 from setuptools import setup, find_packages
 
-install_reqs = ["Django>=1.4", "sixpack-client==1.0.0"]
+install_reqs = ["Django>=1.4", "sixpack-client==1.1.1"]
```
Comment (dlanger) [inline setup.py]: Are you sure it's still `Django>=1.4`, given the migrations you've added?

File: djsixpack/models.py (hunk 8 above)
Comment (4x4notfound) [inline djsixpack/models.py]: yes

File: setup.py (hunk 9 above)
Comment (4x4notfound) [inline setup.py]: I crossed check all the APIs used added and they are available in 1.4...is there anything else I should check?

File: djsixpack/djsixpack.py [hunk 10]
```diff
@@ -73,16 +86,26 @@ def participate(self, force=None, user_agent=None, ip_address=None):
...
+        chosen_alternative = None
+
+        if self.local and not self.server:
+            prefetch = True
 
         try:
-            resp = session.participate(experiment_name, self.alternatives, force)
+            resp = session.participate(experiment_name, self.alternatives, force=force, prefetch=prefetch)
         except RequestException:
             logger.exception("Error while trying to .participate")
```
Comment (squishybear-zz) [inline djsixpack/djsixpack.py:99]: While you're in here, please make the quotation use consistent.

**Comment by squishybear-zz** [comment]
LGTM

File: setup.py (hunk 9 above)
Comment (dlanger) [inline setup.py]: I'm thinking around migrations - if it's an old-style `south` migration, don't we need to add a requirement for `south`?

@ashchristopher, can you confirm how we should be handling migrations to work with both old and new? You've probably done more research into this than anyone else.

Comment (adrianmaurer) [inline setup.py]: Good point....

I ran the migration in accounting with `pip install -e ../sandbox/djang-sixpack` and it picked up and migrated okay. Should we use 1.5 as the base?

**Comment by ashchristopher** [comment]
@dlanger @adrianmaurer - re: South - this could be a bit tricky - grab me first thing in the morning and we can go over it. The short of it is, you will need to make South==1.0.0 a requirement of this package.

**Comment by 4x4notfound** [comment]
right, south isn't included by default in 1.5.

File: README.rst [hunk 11]
```diff
@@ -87,6 +87,27 @@ If something ever goes wrong - a request times out, the ``sixpack`` server disap
...
+
+This feature is particullary useful when you want to pull information on an existing participant, a feature unavailable through Sixpack API.
+
+.. code:: python
+   
+   experiment = ButtonColorTest(local=True)
+
+You may also choose to track only locally by passing in `sixpack=False` to the test constructor.
```
Comment (squishybear-zz) [inline README.rst:102]: This description doesn't match the sample code on line 106.

**Comment by mwarkentin** [comment]
Closing as stale.

**Pull request closed without merging**